codegen/gce/src/lib.rs

```diff
@@ -0,0 +1,248 @@
+use ir::{
+    constraints::{ConstantValue, Operation},
+    AirIR,
+};
+
+pub use air_script_core::{
+    Constant, ConstantType, Expression, Identifier, IndexedTraceAccess, MatrixAccess,
+    NamedTraceAccess, TraceSegment, Variable, VariableType, VectorAccess,
+};
+use std::fmt::Display;
+
+mod error;
+use error::ConstraintEvaluationError;
+
+mod utils;
+
+mod expressions;
+use expressions::ExpressionsHandler;
+
+use std::collections::BTreeMap;
+use std::fs::File;
+use std::io::Write;
+
+/// Holds data for JSON generation
+#[derive(Default, Debug)]
+pub struct CodeGenerator {
+    num_polys: u16,
+    num_variables: usize,
+    constants: Vec<u64>,
+    expressions: Vec<ExpressionJson>,
+    outputs: Vec<usize>,
+}
+
+impl CodeGenerator {
+    pub fn new(ir: &AirIR, extension_degree: u8) -> Result<Self, ConstraintEvaluationError> {
+        // maps indexes in Node vector in AlgebraicGraph and in `expressions` JSON array
+        let mut expressions_map = BTreeMap::new();
+
+        let num_polys = set_num_polys(ir, extension_degree);
+        let num_variables = set_num_variables(ir);
+        let constants = set_constants(ir);
+
+        let mut expressions_handler = ExpressionsHandler::new(ir, &constants, &mut expressions_map);
+
+        let mut expressions = expressions_handler.get_expressions()?;
+        // vector of `expressions` indexes
+        let outputs = expressions_handler.get_outputs(&mut expressions)?;
```

Comment:
Instead of this pattern where you get the expressions and then pass them back in, I would keep them in the `ExpressionsHandler`. I suggest adjusting to a builder pattern. We could rename `ExpressionsHandler` to `GceBuilder`, add a method which handles building the `expressions` and `outputs`, and then add an `into` method that consumes the builder and yields the `expressions` and `outputs`. (This would also reflect the pattern that we'll be moving the IR to in #150)

Something like:

```
    pub fn new(ir: &AirIR, extension_degree: u8) -> Result<Self, ConstraintEvaluationError> {
        let num_polys = set_num_polys(ir, extension_degree);
        let num_variables = set_num_variables(ir);
        let constants = set_constants(ir);

        let builder = GceBuilder::new();
        builder.build(ir, &constants)?;
        let (expressions, outputs) = builder.into_gce();

        Ok(CodeGenerator {
            num_polys,
            num_variables,
            constants,
            expressions,
            outputs,
        })
    }
```